cilium/CFP-30235-xds-adapter.md

```diff
@@ -0,0 +1,213 @@
+universal data plane API. xDS is CNCF-governed and is gradually transitioning
+content from the
+[envoyproxy/data-plane-api](https://github.com/envoyproxy/data-plane-api) repo
+to [cncf/xds](https://github.com/cncf/xds) repo as part of that overall vision.
+
+In parallel to these efforts by the xDS community, GKE is planning to introduce
+xDS as an additional data source for DPv2 configuration. This feels sufficiently
+generic and helpful that it could be something that could be contributed to
+upstream Cilium. This could be particularly useful for at least two common use
+cases:
+
+1. Supporting Services and Endpoints from outside of the local cluster.
+1. Supporting advanced routing techniques, such as topology aware routing.
+
+Although it is out of scope for this specific CFP to provide complete solutions
+for either of these use cases, it will demonstrate the benefits of having an
+xDS adapter when developing a solution for either of these use cases.
+
+### Use Cases
+
+#### Topology Aware Routing
+
+As the author of both EndpointSlices and the existing Topology Aware Routing
+approach in Kubernetes, I feel like I’m well equipped to discuss the limitations
+of both. What most people want from Topology Aware Routing is essentially to
+fill the local zone to capacity and then spillover to the next closest set of
+endpoints with capacity. Unfortunately that is very difficult to achieve with
+the existing Kubernetes tooling. There are a few key problems here:
+
+1. Kubernetes has no feedback loop to its dataplane and therefore no reliable
+   way to understand if endpoints have reached capacity.
+2. Even if there were a feedback loop, spillover to other zones benefits from
+   some form of centralized orchestration to avoid a thundering herd problem.
+3. All changes to endpoint routing currently have to go through Services or
+   EndpointSlices, and by extension, the Kubernetes API Server. It would be very
+   expensive to implement incremental weighting/spillover adjustments with that
+   overhead. For example, adjusting the weight of any individual endpoint would
+   require a write to the EndpointSlice API which would then need to be
+   distributed to all consumers of that API. That can be problematic from a
+   scalability perspective when there are frequent updates paired with a large
+   number of endpoints and/or nodes.
+
+I believe that xDS is uniquely positioned to address these limitations.
+[LRS](https://www.envoyproxy.io/docs/envoy/latest/api-v3/service/load_stats/v3/lrs.proto)
+(Load Reporting Service) provides a straightforward way to provide load reports
+to a centralized control plane. Similarly, [Delta
+xDS](https://www.envoyproxy.io/docs/envoy/latest/configuration/overview/xds_api#delta-endpoints)
+enables xDS to distribute incremental updates that can be smaller in scope than
+comparable Kubernetes APIs. All of this could be combined to avoid adding
+additional load on the Kubernetes API Server.
```

Comment:
I agree that LRS is designed to do exactly this, but implementing incremental xDS is _not_ straightforward. Unless your control plane is very carefully designed, it's nearly impossible (which is why many simpler Envoy control planes don't do it.)

To be clear, the hard part is on whoever's building the control plane, not inside the Cilium Agent. But given that we'll need an open-source thing to test with at the very least, the engineering cost of building this shouldn't be underestimated.

Reply:
Discussing *not* running this in production gives me a bit of concern. How will Cilium users benefit from this functionality if the reference implementation isn't built with production in mind?

Reply:
To clarify my previous comment, I am not suggesting we should build an experimental control plane. Instead, we should agree on the scope of the work as an initial deliverable.  Building a production level control plane that handles all the edge cases can be challenging. There should definitely be a reference implementation that is usable outside of tests.

Reply:
I think the minimum bar here in terms of initial development would be for a OSS/CNCF reference xDS control plane that is sufficiently reliable to run for e2e testing and development and also at least one that is production ready. Those will probably be the same project, but if we are successful at building a good ecosystem here, it's possible that there will be multiple OSS production-ready options and we won't need the reference/testing implementation to also be production ready. Our e2e tests here could become something resembling conformance tests, but for control plane implementations.